README.md
# Introduction

The Skytraq-datalogger project is cloned into the github for further preservation.

This project is in the `works for me` state, i will not maintain this project.
Please see [http://code.google.com/p/skytraq-datalogger/] for the original resource.

The project is under GPLv2 License, original work by some real nice dudes, thank you!

# Original README from the project page

Recently I bought a Skytraq DL-65 bluetooth GPS data logger (55 Euro). The provided software to read tracks from the device is a Windows program the "GPS Photo Tagger" from iTravel-Tech. But since I do not use Windows I needed a software for Linux to download track and configure the device.

When you connect the data logger via USB you can see that it uses a PL2303 USB-serial converter. Without any further software you can dump the NMEA data stream: cat /dev/ttyUSB0.

My device contains a SkyTraq Venus 6 GPS receiver. Older data loggers contain a Venus 5 GPS receiver. Those should also work with my program. The SkyTraq Venus uses a binary protocol to communicate with the host. 


debian/changelog
skytraq-datalogger (0.4-1) unstable; urgency=low

  * 2009-08-26: GPX output
                support for devices that have a button to tag waypoints

 -- Jesper Zedlitz <jesper@zedlitz.de> Wed, 26 Aug 2009 12:49:24 +0200

skytraq-datalogger (0.3-1) unstable; urgency=low

  * 2008-10-23: read status of AGPS data
                the baud-rate detection routine also uses NMEA output

 -- Jesper Zedlitz <jesper@zedlitz.de> Thu, 23 Oct 2008 12:38:03 +0200



test.c
#include "agps-download.h"
#include "datalogger.h"
#include "lowlevel.h"
#include <stdlib.h>

int main(int argc, char *argv[])  {
    int fd;
    int speed;
    agps_data data;

    fd = open_port("/dev/ttyUSB0");
    speed =   skytraq_determine_speed(fd);

    printf("opened device with %d bit/s\n", speed);

    if ( skytraq_download_agps_data(&data) ) {
        printf("sending data...\n");
        skytraq_send_agps_data( fd, &data );
        free(data.memory);
    }


    close(fd);

    return 0;
}



main.c
/*

    Control program for SkyTraq GPS data logger.

    Copyright (C) 2008  Jesper Zedlitz, jesper@zedlitz.de

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111 USA

 */

#include "datalogger.h"
#include "lowlevel.h"

enum{ NO_ACTION, ACTION_INFO, ACTION_DELETE, ACTION_DUMP, ACTION_CONFIG,
      ACTION_SET_SPEED, ACTION_OUTPUT_OFF, ACTION_OUTPUT_NMEA, ACTION_OUTPUT_BINARY
    };

#define verbose(fmt, args...) fprintf(stderr, fmt"\n", ##args)

int main(int argc, char *argv[])  {
    int fd, i, baud_rate = 0, action=NO_ACTION, serial_speed=0;
    int min_time=-1, max_time=-1, min_dist=-1, max_dist=-1, min_speed=-1, max_speed=-1, enable=0, disable=0;
    int mode_fifo = 0, mode_stop = 0;
    int permanent = 0;
    int success;
    char* device = "/dev/ttyUSB0";

    for ( i = 1; i < argc; i++ ) {
        if (!strcmp(argv[i], "--info" ) ) {
            action = ACTION_INFO;
        } else if ( !strcmp(argv[i], "--delete" ) ) {
            action = ACTION_DELETE;
        } else if ( !strcmp(argv[i], "--dump" ) ) {
            action = ACTION_DUMP;
        } else if ( !strcmp(argv[i], "--set-config" ) ) {
            action = ACTION_CONFIG;
        } else if ( !strcmp(argv[i], "--set-baud-rate" ) ) {
            action = ACTION_SET_SPEED;
	    if ( argc>i+1) serial_speed = atoi(argv[++i]);
        } else if ( !strcmp(argv[i], "--device" ) ) {
            if ( argc>i+1) device = argv[++i];
        } else if ( !strcmp(argv[i], "--time" ) ) {
            if ( argc>i+1) min_time = atoi(argv[++i]);
        } else if ( !strcmp(argv[i], "--max-time" ) ) {
            if ( argc>i+1) max_time = atoi(argv[++i]);
        } else if ( !strcmp(argv[i], "--dist" ) ) {
            if ( argc>i+1) min_dist = atoi(argv[++i]);
        } else if ( !strcmp(argv[i], "--max-dist" ) ) {
            if ( argc>i+1) max_dist = atoi(argv[++i]);
        } else if ( !strcmp(argv[i], "--speed" ) ) {
            if ( argc>i+1) min_speed = atoi(argv[++i]);
        } else if ( !strcmp(argv[i], "--max-speed" ) ) {
            if ( argc>i+1) max_speed = atoi(argv[++i]);
        } else if ( !strcmp(argv[i], "--enable-log" ) ) {
            enable = 1;
        } else if ( !strcmp(argv[i], "--disable-log" ) ) {
            disable = 1;
        } else if ( !strcmp(argv[i], "--mode-fifo" ) ) {
            mode_fifo = 1;
        } else if ( !strcmp(argv[i], "--mode-stop" ) ) {
            mode_stop = 1;
        } else if ( !strcmp(argv[i], "--permanent" ) ) {
            permanent = 1;
        } else if ( !strcmp(argv[i], "--set-output-off" ) ) {
            action = ACTION_OUTPUT_OFF;
        } else if ( !strcmp(argv[i], "--set-output-nmea" ) ) {
            action = ACTION_OUTPUT_NMEA;
        } else if ( !strcmp(argv[i], "--set-output-bin" ) ) {
            action = ACTION_OUTPUT_BINARY;
        } else if ( !strcmp(argv[i], "--baud-rate" ) ) {
	    if ( argc>i+1) baud_rate = atoi(argv[++i]);
	}
    }

    if ( action == NO_ACTION ) {
        fprintf(stderr, "USAGE: %s <OPTIONS> ACTION \n", argv[0]);
        fprintf(stderr, " ACTION is one of:\n");
        fprintf(stderr, "  --info             get information about software version and configuration\n");
        fprintf(stderr, "  --delete           delete all track lists from the data logger\n");
        fprintf(stderr, "  --dump             dump track lists to STDOUT\n");
        fprintf(stderr, "  --set-config       change configuration of the data logger\n");
        fprintf(stderr, "  --set-baud-rate    configure speed of the device's serial port\n");
        fprintf(stderr, "  --set-output-off   disable output for GPS data\n");
        fprintf(stderr, "  --set-output-nmea  enable output for GPS data in NMEA format\n");
        fprintf(stderr, "  --set-output-bin   enable output for GPS data in binary format\n");
        fprintf(stderr, " OPTIONS:\n");
        fprintf(stderr, "  --device <DEV>        name of the device, default is /dev/ttyUSB0\n");
        fprintf(stderr, "  --permanent           write serial port speed to FLASH\n");
	fprintf(stderr, "  --baud-rate           set baud-rate manually\n");
        fprintf(stderr, " OPTIONS for configuration:\n");
        fprintf(stderr, "  --time <SECONDS>      log every <SECONDS> seconds\n");
        fprintf(stderr, "  --max-time <SECONDS>  \n");
        fprintf(stderr, "  --dist <METERS>       log every <METERS> meters\n");
        fprintf(stderr, "  --max-dist <METERS>   \n");
        fprintf(stderr, "  --speed <KMPH>        only log if faster than <KMPH> km/h\n");
        fprintf(stderr, "  --max-speed <KMPH>   \n");
        fprintf(stderr, "  --enable-log          turn on logging\n");
        fprintf(stderr, "  --disable-log         turn off logging\n");
        fprintf(stderr, "  --mode-fifo           overwrite oldest entries when no space is left\n");
        fprintf(stderr, "  --mode-stop           stop logging when no space is left\n");
        return 2;
    }

    fd = open_port(device);
    if ( fd == -1 ) {
        fprintf(stderr,"Failed to open device %s\n", device);
        return 1;
    }

    /* detect device and speed */
    if( baud_rate == 0 ) {
	    baud_rate = skytraq_determine_speed(fd);
	    if ( baud_rate == 0 ) {
	        fprintf(stderr,"Could not find data logger at port %s\n", device);
	        return 1;
	    }
    }

    /* get status and config from GPS data logger */
    skytraq_config* info = malloc( sizeof(skytraq_config));
    success = skytraq_read_datalogger_config(fd,info);
    
    if( success != SUCCESS ) {
    	fprintf(stderr, "No response from datalogger.\n");
	return 1;
    }

    if ( action == ACTION_INFO ) {
        int agps_days, agps_hours;
        skytraq_read_software_version(fd);
	skytraq_read_agps_status(fd, info);
	
	agps_days = info->agps_hours_left / 24;
	agps_hours = info->agps_hours_left % 24;
	
        printf("log_wr_ptr:      %ld\n",info->log_wr_ptr );
        printf("total sectors:   %d\n", info->total_sectors);
        printf("sectors left:    %d\n", info->sectors_left);
        printf("max time:        %ld s\n", info->max_time);
        printf("min time:        %ld s\n", info->min_time);
        printf("max distance:    %ld m\n", info->max_distance);
        printf("min distance:    %ld m\n",info->min_distance );
        printf("max speed:       %ld km/h\n", info->max_speed);
        printf("min speed:       %ld km/h\n", info->min_speed);
        printf("datalog enable:  %d\n",info->datalog_enable );
        printf("log fifo mode:   %d\n", info->log_fifo_mode);
	printf("AGPS enabled:    %d\n", info->agps_enabled);
	printf("AGPS data left:  ");
	if( agps_days == 1 ) {
	  printf("1 day ");
	} else if( agps_days > 0 ) {
	  printf("%d days ", agps_days);
	}
	if( agps_hours == 1 ) {
	  printf("1 hour");
	} else if( agps_hours > 0 ) {
	  printf("%d hours ", agps_hours);
	}
	printf("\n");
	printf("baud-rate:       %d bps\n", baud_rate);
    } else if ( action == ACTION_DELETE ) {
        skytraq_clear_datalog(fd);
    } else if ( action == ACTION_DUMP ) {
        int used_sectors;

        used_sectors = info->total_sectors - info->sectors_left + 1;

	/* print GPX header */
	printf("<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n");
	printf("<gpx xmlns=\"http://www.topografix.com/GPX/1/0\" creator=\"skytraq-datalogger\" version=\"1.0\">\n");
	printf("<trk>\n<trkseg>\n");
	
        for ( i = 0; i< used_sectors ; i++ ) {
            int len, retries_left = 3;
            gbuint8* buf = malloc(4100);
            len= skytraq_read_datalog_sector(fd,i,buf);
            while ( len == -1 && retries_left > 3 ) {
                /* retry to read the sector */
                len= skytraq_read_datalog_sector(fd,i,buf);
                retries_left--;
            }
            process_buffer(buf,len);
            free(buf);

            sleep(1);
            /* close(fd);
             fd = open("/dev/ttyUSB0",O_RDWR); */
        }
	
	printf("</trkseg>\n</trk>\n</gpx>\n");
	
    } else if ( action  == ACTION_CONFIG ) {
        if ( min_time > -1 ) info->min_time = min_time;
        if ( max_time > -1 ) info->max_time = max_time;
        if ( min_dist > -1 ) info->min_distance = min_dist;
        if ( max_dist > -1 ) info->max_distance = max_dist;
        if ( min_speed > -1 ) info->min_speed = min_speed;
        if ( max_speed > -1 ) info->max_speed = max_speed;
        if ( disable ) info->datalog_enable = 0;
        if ( enable ) info->datalog_enable = 1;
        if ( mode_stop ) info->log_fifo_mode = 0;
        if ( mode_fifo ) info->log_fifo_mode = 1;
        skytraq_write_datalogger_config(fd,info);
    } else if( action == ACTION_SET_SPEED ) {
    	unsigned requested_speed = skytraq_mkspeed( serial_speed );
	if( requested_speed != ERROR ) {
		skytraq_set_serial_speed(fd,requested_speed,permanent);
	} else {
	   fprintf( stderr, "unknown speed %d\n", serial_speed);
	   return 1;
	}
    } else if( action == ACTION_OUTPUT_OFF ) {
    	skytraq_output_disable(fd);
    } else if( action == ACTION_OUTPUT_NMEA ) {
    	skytraq_output_enable_nmea(fd);    
    } else if( action == ACTION_OUTPUT_BINARY ) {
       	skytraq_output_enable_binary(fd);
    }
    free(info);
    close(fd);

    return 0;
}


datalogger.h
/*

    Control program for SkyTraq GPS data logger.

    Copyright (C) 2008  Jesper Zedlitz, jesper@zedlitz.de

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111 USA

 */

#ifndef datalogger_h
#define datalogger_h

#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <termios.h>
#include <math.h>
#include <unistd.h>
#include <time.h>
#include <sys/time.h>
#include "agps-download.h"

enum{ERROR=-1, SUCCESS};
enum{NACK=-1, ACK};

#define gbuint8 unsigned char
#define gbuint32 unsigned long
#define gbuint16 unsigned int

#ifdef DEBUG_ALL
#undef DEBUG
#define DEBUG(fmt, args...) fprintf(stderr, fmt, ##args)
#else
#define DEBUG(fmt, args...)
#endif

#define SKYTRAQ_SPEED_4800      0
#define SKYTRAQ_SPEED_9600      1
#define SKYTRAQ_SPEED_19200     2
#define SKYTRAQ_SPEED_38400     3
#define SKYTRAQ_SPEED_57600     4
#define SKYTRAQ_SPEED_115200    5

typedef struct SkyTraqPackage {
    gbuint8	length;
    gbuint8* 	data;
    gbuint8 	checksum;
} SkyTraqPackage;

typedef struct skytraq_config {
    gbuint32    log_wr_ptr;
    gbuint16    sectors_left;
    gbuint16    total_sectors;
    gbuint32	max_time;
    gbuint32	min_time;
    gbuint32	max_distance;
    gbuint32	min_distance;
    gbuint32	max_speed;
    gbuint32	min_speed;
    gbuint8	datalog_enable;
    gbuint8	log_fifo_mode;
    gbuint8	agps_enabled;
    unsigned	agps_hours_left;
} skytraq_config;

int skytraq_read_software_version( int fd);
int skytraq_read_datalogger_config( int fd, skytraq_config* config);
int skytraq_read_datalog_sector( int fd, gbuint8 sector, gbuint8* buffer );
void skytraq_clear_datalog( int fd);
void skytraq_write_datalogger_config( int fd, skytraq_config* config);
void process_buffer(gbuint8* buffer, int length);
int skytraq_determine_speed( int fd) ;
unsigned skytraq_mkspeed(unsigned br);
int skytraq_set_serial_speed( int fd, int speed, int permanent);
void skytraq_read_agps_status(int fd, skytraq_config* config);
int skytraq_output_disable( int fd );
int skytraq_output_enable_nmea( int fd );
int skytraq_output_enable_binary( int fd );
int skytraq_send_agps_data( int fd, agps_data* data );

#endif


datalogger.c
/*

    Control program for SkyTraq GPS data logger.

    Copyright (C) 2008  Jesper Zedlitz, jesper@zedlitz.de

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111 USA

 */
#include "datalogger.h"
#include "lowlevel.h"
#include "agps-download.h"

#define SKYTRAQ_COMMAND_SYSTEM_RESTART           0x01
#define SKYTRAQ_COMMAND_QUERY_SOFTWARE_VERSION   0x02
#define SKYTRAQ_COMMAND_QUERY_SOFTWARE_CRC       0x03
#define SKYTRAQ_COMMAND_SET_FACTORY_DEFAULTS     0x04
#define SKYTRAQ_COMMAND_CONFIGURE_SERIAL_PORT    0x05
#define SKYTRAQ_COMMAND_CONFIGURE_NMEA_MESSAGE   0x08
#define SKYTRAQ_COMMAND_CONFIGURE_MESSAGE_TYPE   0x09
#define SKYTRAQ_COMMAND_GET_CONFIG               0x17
#define SKYTRAQ_COMMAND_WRITE_CONFIG             0x18
#define SKYTRAQ_COMMAND_ERASE                    0x19
#define SKYTRAQ_COMMAND_READ_SECTOR              0x1b
#define SKYTRAQ_COMMAND_GET_EPHERMERIS           0x30
#define SKYTRAQ_COMMAND_SET_EPHEMERIS            0x31
#define SKYTRAQ_COMMAND_READ_AGPS_STATUS         0x34
#define SKYTRAQ_COMMAND_SEND_AGPS_DATA           0x35
#define SKYTRAQ_RESPONSE_SOFTWARE_VERSION        0x80
#define SKYTRAQ_RESPONSE_SOFTWARE_CRC            0x81
#define SKYTRAQ_RESPONSE_ACK                     0x83
#define SKYTRAQ_RESPONSE_NACK                    0x84
#define SKYTRAQ_RESPONSE_EPHEMERIS_DATA          0xb1

#define TIMEOUT  2000l

int skytraq_read_software_version( int fd) {
    int result = ERROR;
    SkyTraqPackage* request = skytraq_new_package(2);
    request->data[0] = SKYTRAQ_COMMAND_QUERY_SOFTWARE_VERSION;
    request->data[1] = 1;
    if ( ACK == skytraq_write_package_with_response(fd,request,TIMEOUT)) {
        SkyTraqPackage* response = skytraq_read_next_package(fd,TIMEOUT);
        if ( response != NULL) {
            skytraq_dump_package(response);

            printf("kernel version: %d.%d.%d -- ODM version: %d.%d.%d -- revision: 20%02d-%02d-%02d\n",
                   response->data[3],response->data[4],response->data[5],
                   response->data[7],response->data[8],response->data[9],
                   response->data[11],response->data[12],response->data[13]);
            skytraq_free_package(response);
            result = SUCCESS;
        }
    }

    skytraq_free_package(request);

    return result;
}

void skytraq_write_datalogger_config( int fd, skytraq_config* config) {
    SkyTraqPackage* request = skytraq_new_package(27);
    request->data[0] = SKYTRAQ_COMMAND_WRITE_CONFIG;
    request->data[1] = (config->max_time>>24) & 0xff;
    request->data[2] = (config->max_time>>16) & 0xff;
    request->data[3] = (config->max_time>>8) & 0xff;
    request->data[4] = config->max_time & 0xff;
    request->data[5] = (config->min_time>>24) & 0xff;
    request->data[6] = (config->min_time>>16) & 0xff;
    request->data[7] = (config->min_time>>8) & 0xff;
    request->data[8] = config->min_time & 0xff;
    request->data[9] = 00;
    request->data[10] = 00;
    request->data[11] = 03;
    request->data[12] = 0xE8;
    request->data[13] = 00;
    request->data[14] = 00;
    request->data[15] = 00;  /* distance HIGH */
    request->data[16] = config->min_distance & 0xff;
    request->data[17] = 00;
    request->data[18] = 00;
    request->data[19] = 03;
    request->data[20] = 0xE8;
    request->data[21] = 00;
    request->data[22] = 00;
    request->data[23] = (config->min_speed>>8) & 0xff;
    request->data[24] = config->min_speed & 0xff;
    request->data[25] = config->datalog_enable;
    request->data[26] = config->log_fifo_mode;
    skytraq_write_package_with_response(fd,request,TIMEOUT);
    skytraq_free_package(request);
}

unsigned uint32_from_buffer( gbuint8* buffer, int offset ) {
    return buffer[offset] | (buffer[offset+1]<<8) |
           (buffer[offset+2]<<16) | (buffer[offset+3]<<24);
}

unsigned uint16_from_buffer( gbuint8* buffer, int offset ) {
    return buffer[offset]| (buffer[offset+1]<<8);
}

int skytraq_read_datalogger_config( int fd, skytraq_config* config ) {
    int result = ERROR;
    SkyTraqPackage* request = skytraq_new_package(1);
    request->data[0] = SKYTRAQ_COMMAND_GET_CONFIG;
    if ( ACK == skytraq_write_package_with_response(fd,request,TIMEOUT)) {
        SkyTraqPackage* response = skytraq_read_next_package(fd,TIMEOUT);
        if ( response != NULL) {
            skytraq_dump_package(response);

            config->log_wr_ptr = uint32_from_buffer(response->data,1);
            config->sectors_left = uint16_from_buffer(response->data,5);
            config->total_sectors = uint16_from_buffer(response->data,7);
            config->max_time = uint32_from_buffer(response->data,9);
            config->min_time  = uint32_from_buffer(response->data,13);
            config->max_distance   = uint32_from_buffer(response->data,17);
            config->min_distance  = uint32_from_buffer(response->data,21);
            config->max_speed = uint32_from_buffer(response->data,25);
            config->min_speed = uint32_from_buffer(response->data,29);
            config->datalog_enable = response->data[33];
            config->log_fifo_mode = response->data[34];

            skytraq_free_package(response);
            result = SUCCESS;
        }
    }

    skytraq_free_package(request);
    return result;
}

void skytraq_clear_datalog( int fd) {
    SkyTraqPackage* request= skytraq_new_package(1);
    request->data[0] = 0x19;
    skytraq_write_package_with_response(fd,request,TIMEOUT);
    skytraq_free_package(request);
}

/**
 * Returns number of bytes read.
 */
int skytraq_read_datalog_sector( int fd, gbuint8 sector, gbuint8* buffer ) {
    if ( buffer != NULL ) {
        SkyTraqPackage* request= skytraq_new_package(2);;
        request->data[0] = 0x1b;
        request->data[1] = sector;
        if ( ACK == skytraq_write_package_with_response(fd,request,TIMEOUT)) {
            int i,len, count = 0;
            gbuint8 c, lastByte1, lastByte2,cs,checksum;

            DEBUG("START READING DATA\n");

            cs = 0;
            len = read_with_timeout( fd, &c, 1,TIMEOUT);
            while ( len > 0 ) {
                if ( (lastByte2 == 'E') && (lastByte1 == 'N') && (c == 'D')) {
                    /* remove last two bytes from checksum */
                    cs = cs ^'N';
                    cs = cs ^'E';

                    DEBUG("DONE\n");
                    break;
                }


                if ( count >=2 ) {
                    buffer[count-2] = lastByte2;
                }

                cs = cs ^ c;

                lastByte2 = lastByte1;
                lastByte1 = c;
                count++;
                len = read_with_timeout( fd, &c, 1,TIMEOUT);
            }


            /* remaining characters after data block */
            for ( i=0; i<10; i++)
                read_with_timeout( fd, &c, 1, TIMEOUT);

            read_with_timeout( fd, &checksum, 1, TIMEOUT);

            for ( i=0; i<5; i++)
                read_with_timeout( fd, &c, 1, TIMEOUT);


            if ( cs == checksum )
                return count-2;
            else {
                fprintf(stderr, "wrong checksum for sector %d\n", sector);
                return -1;
            }
        }

        skytraq_free_package(request);
    }

    return -1;
}


unsigned skytraq_mkspeed(unsigned br) {
    switch (br) {
    case   4800:
        return  SKYTRAQ_SPEED_4800;
    case   9600:
        return  SKYTRAQ_SPEED_9600;
    case  19200:
        return  SKYTRAQ_SPEED_19200;
    case  38400:
        return  SKYTRAQ_SPEED_38400;
    case  57600:
        return  SKYTRAQ_SPEED_57600;
    case 115200:
        return SKYTRAQ_SPEED_115200;
    default:
        return ERROR;
    }
}

/**
  * Set the speed used on the serial line.
  */
int skytraq_set_serial_speed( int fd, int speed, int permanent) {
    SkyTraqPackage* request= skytraq_new_package(4);
    request->data[0] = SKYTRAQ_COMMAND_CONFIGURE_SERIAL_PORT;
    request->data[1] = 0;
    request->data[2] = speed;
    request->data[3] = permanent & 1;
    if ( ACK != skytraq_write_package_with_response(fd, request,TIMEOUT) ) {
        fprintf( stderr, "setting line speed FAILED\n");
        return 1;
    }
    skytraq_free_package(request);
    return 0;
}


/*unsigned baud_rates[] = { 115200, 9600,57600 ,1200,2400,4800,19200, 38400 };*/
unsigned baud_rates[] = { 9600,115200, 57600, 38400, 19200, 9600, 4800, 2400, 1200 };

int contains( gbuint8* haystack, int h_length, gbuint8* needle, int n_length) {
    int offset = 0;
    int i;
    while ( offset < h_length-n_length ) {
        int found = 0;
        for ( i= 0; i<n_length ; i++) {
            if ( haystack[offset+i] == needle[i] ) {
                found++;
            }
        }
        if ( found == n_length ) {
            DEBUG("found needle at offset %d\n", offset);
            return 1;
        }
        offset++;
    }

    return 0;
}



int skytraq_determine_speed( int fd) {
    gbuint8 request[] = { 0xA0, 0xA1, 0x00, 0x02, 0x02, 0x01, 0x03, 0x0D, 0x0A };
    gbuint8 expected_response[] = { 0xA0, 0xA1, 0x00, 0x02, 0x83, 0x02, 0x81, 0x0D, 0x0A};
    int buf_size=4000;
    gbuint8* buffer = malloc(buf_size);
    gbuint8 expected_nmea_responce[] = "$GP";
    int nmea_detection_succeded=0;
    int binary_detection_succeded=0;
    int len,i;
    for ( i = 0; i< (sizeof(baud_rates)/sizeof(int)); i++) {
        DEBUG("testing for %d baud-rate\n", baud_rates[i]);
        set_port_speed(fd, baud_rates[i]);
        /* If baud rate is right we will receive some NMEA data first
         * otherwise at least we will let device enough characters
         * to respond to previous requests */
        len = read_with_timeout(fd, buffer, buf_size,200);

#ifdef DEBUG_ALL
DEBUG("len=%d\n",len);
        if (len < (buf_size-1))
            buffer[len]=0; /* in C strings must be terminated by 0 */
        else
            buffer[buf_size-1]=0; /* we truncate one symbol from the buffer */
        DEBUG("current ascii output: \n");
        DEBUG("received %d bytes\n", len);
        DEBUG("buffer as ascii contains: %s\n", buffer);
#endif
        if ( contains( buffer, len, expected_nmea_responce, 3)) {
            nmea_detection_succeded=1;
            DEBUG("NMEA detection succeeded\n");
            DEBUG("detected speed is %d\n", baud_rates[i]);
        } else {
            /* try binary detection */

            /* send "QUERY SOFTWARE VERSION" command to GPS unit */
            write(fd, request, 9);

            len = read_with_timeout(fd, buffer, buf_size,200);
#ifdef DEBUG_ALL
DEBUG("len=%d\n",len);
            if (len < (buf_size-1))
                buffer[len]=0; /* in C strings must be terminated by 0 */
            else
                buffer[buf_size-1]=0; /* we truncate one symbol from the buffer */
            DEBUG("after binary request output: \n");
            DEBUG("received %d bytes\n", len);
            DEBUG("buffer as ascii contains: %s\n", buffer);
#endif

            if ( contains( buffer, len, expected_response, 3)) {
                binary_detection_succeded=1;
                DEBUG("binary detection succeeded\n");
                DEBUG("detected speed is %d\n", baud_rates[i]);
            }
        }

        if (binary_detection_succeded || nmea_detection_succeded) {
            free(buffer);
            return baud_rates[i];
        }
    }

    free(buffer);
    set_port_speed(fd, 38400); /* since we unlucky we set port speed to something high again */
    return 0;
}

/**
  * Determine how many hours of AGPS data is available on the device
  */
void skytraq_read_agps_status(int fd, skytraq_config* config)  {
    struct tm* date;
    time_t now;
    config->agps_hours_left = 0;
    config->agps_enabled = 0;

    time(&now);
    date = gmtime(&now);

    SkyTraqPackage* request = skytraq_new_package(8);
    request->data[0] = SKYTRAQ_COMMAND_READ_AGPS_STATUS;
    request->data[1] = ((date->tm_year + 1900)>>8) & 0xff;
    request->data[2] = (date->tm_year + 1900)&0xff;
    request->data[3] = date->tm_mon + 1;
    request->data[4] = date->tm_mday ;
    request->data[5] = date->tm_hour;
    request->data[6] = date->tm_min;
    request->data[7] = date->tm_sec;

    if ( ACK == skytraq_write_package_with_response(fd,request,TIMEOUT)) {
        SkyTraqPackage* response = skytraq_read_next_package(fd,TIMEOUT);
        if ( response != NULL) {
            skytraq_dump_package(response);

            config->agps_hours_left = uint16_from_buffer(response->data,1);
            config->agps_enabled    = response->data[3];

            skytraq_free_package(response);
        }
    }

    skytraq_free_package(request);
}

int skytraq_output_disable( int fd ) {
    int success;
    SkyTraqPackage* request = skytraq_new_package(3);
    request->data[0] = SKYTRAQ_COMMAND_CONFIGURE_MESSAGE_TYPE;
    request->data[1] = 0;
    request->data[2] = 0;  /* 1 = permanent */

    success = skytraq_write_package_with_response(fd,request,TIMEOUT);
    skytraq_free_package(request);
    return success;
}

int skytraq_output_enable_nmea( int fd ) {
    int success;
    SkyTraqPackage* request = skytraq_new_package(3);
    request->data[0] = SKYTRAQ_COMMAND_CONFIGURE_MESSAGE_TYPE;
    request->data[1] = 1;
    request->data[2] = 0;  /* 1 = permanent */

    success = skytraq_write_package_with_response(fd,request,TIMEOUT);
    skytraq_free_package(request);
    return success;
}

int skytraq_output_enable_binary( int fd ) {
    int success;
    SkyTraqPackage* request = skytraq_new_package(3);
    request->data[0] = SKYTRAQ_COMMAND_CONFIGURE_MESSAGE_TYPE;
    request->data[1] = 2;
    request->data[2] = 0;  /* 1 = permanent */

    success = skytraq_write_package_with_response(fd,request,TIMEOUT);
    skytraq_free_package(request);
    return success;
}

int skytraq_read_ok( int fd ) {
   gbuint8 buf[50];
   int len;
   
   len = read_string(fd, (gbuint8*)&buf, 50, TIMEOUT );
   
   if( len > 0 ) {
      printf("response from GPS device: %s\n", buf);
      return strncmp( "OK", (char*)&buf, 50) == 0;
   }
   
   return 2;
}

/**
  *  Send block of bytes and wait for OK\0
  */
int skytraq_send_agps_data_block( int fd, gbuint8* data, unsigned block_size ) {
   printf("sending %d bytes from position %p\n", block_size, data);
   write_buffer(fd, data, block_size);
   return skytraq_read_ok( fd );
}


int skytraq_send_agps_data( int fd, agps_data* data  ) {
    gbuint8 info_string[100];
    char buf[4];
    int len;
    unsigned offset  = 0;
    unsigned data_left = data->size;
   
    SkyTraqPackage* request = skytraq_new_package(1);
    request->data[0] = SKYTRAQ_COMMAND_SEND_AGPS_DATA;
    skytraq_write_package_with_response(fd,request,TIMEOUT);
    skytraq_free_package(request);

printf("got ACK\n");

    /* start the transmission */    
    len = snprintf( (char*)info_string, 100, "BINSIZE = %d Checksum = %d Checksumb = %d ", data->size, data->checksumA, data->checksumB );
    info_string[len] = 0;
    printf("%s\n", info_string);
    
    write_buffer(fd, info_string, len+1);
    
printf("waiting for OK\n");    
    if( ! skytraq_read_ok( fd ) ) {
       return ERROR;
    }

    int block_size = 128;
    while( data_left >= block_size ) {
       printf("data left: %d\n", data_left);
       if( skytraq_send_agps_data_block(fd, data->memory + offset , block_size) == 0 ) {
         return ERROR;
       }

       data_left -= block_size;
       offset += block_size;
    }
    
    /* send last block */
    printf("sending last %d bytes\n", data_left);
    skytraq_send_agps_data_block(fd, data->memory + offset , data_left);
    
    read_with_timeout( fd, &buf, 4, TIMEOUT);
        
    return buf[0] == 'E' && buf[1] == 'N' && buf[2] == 'D'&& buf[3] == 0;
}
